lib/api/src/externals/function.rs

```diff
@@ -1022,18 +1025,22 @@ mod inner {
                     /// This is a function that wraps the real host
                     /// function. Its address will be used inside the
                     /// runtime.
-                    extern fn func_wrapper<$( $x, )* Rets, RetsAsResult, Env, Func>( env: &mut Env, $( $x: $x::Native, )* ) -> Rets::CStruct
+                    extern fn func_wrapper<$( $x, )* Rets, RetsAsResult, Env, Func>( env: &std::sync::Mutex<Env>, $( $x: $x::Native, )* ) -> Rets::CStruct
                     where
                         $( $x: FromToNativeWasmType, )*
                         Rets: WasmTypeList,
                         RetsAsResult: IntoResult<Rets>,
                         Env: Sized,
                         Func: Fn(&mut Env, $( $x ),* ) -> RetsAsResult + 'static
                     {
+                        use std::borrow::BorrowMut;
+
                         let func: &Func = unsafe { &*(&() as *const () as *const Func) };
+                        let mut env_guard = env.lock().unwrap();
```

Comment:
I'm not really a fan of this. This is a clear performance penalty that we are now incurring for all function calls.

Reply:
We should measure this penalty, because I don't see a lot more options to solve this sync problem.

Reply:
I'm completely in favor of measurement, but I'm also sympathetic to wanting a fast path for singlethreaded code which will likely be the majority of uses (at least in the short-term).

Also, I agree that yeah, this change is nice because it just works without much changes.  I wrote up a design doc about solutions to this problem and I'll ping people about it next week: I think it's probable that we can make a more complex solution to this problem, but it will require breaking changes to our API or duplicating our API (singlethreaded version vs general version).  I also talked about some more exotic options with Nick in regards to things like updating vtables at runtime to dynamically synchronize code when it becomes multithreaded but I don't think we're realistically going to be trying anything fancy any time soon and also I haven't thought of any fancy solutions that aren't a bad trade off in terms of complexity.